### PR TITLE
Cleaning comAssumption.ml

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** cleanup

This PR is a step towards merging the different paths for `Context` and `Variables` (and their variants).

It includes
- relying on a generic `interp_named_context_evars` to interpret contexts
- checking that evars are resolved the same way
- a new kind of declaration kind for `x:=t` in `Context`, namely `LetContext`
- experimenting removing the special case of `Context` outside sections for `Module Type` (finally done in #18254)
- a few more sanity checks
- coqdoc location for `Variables` like it was for `Context`
- a step towards possibly supporting `Program Context`

To deal with the residual differences of interpretation between `Variable`/`Parameter` and `Context`, we introduce the following parameterization:
- `Variable`/`Parameter` call `interp_named_context_evars` with the share flag on while `Context` calls it with the share flag off. For instance, `Variables A B : Type` shares the universe level but `Context (A B : Type)` introduces two universe levels. See also #13477.
- Deactivation of implicit arguments in `Context` is done by calling `interp_named_context_evars` with implicit mode off.
- `Context` tries to declare variables in classes as instances but not `Variable`/`Parameter`.

Notes:
- Implicit arguments in declarations show the following limits:
  ```coq
  Section A.
  Set Implicit Arguments.
  Variables (f1 : _ )                     (x1 : f1 0 (eq_refl 0) = 0).
  Variables (f2 : forall n:nat, n=n->nat) (x2 : f2   (eq_refl 0) = 0).
  About f1.
  About f2. (* same type and implicit arguments as f1 *)
  ```
- What to do with the discrepancy between the first and the other monomorphic axioms of a bunch of declarations (see #13477)
- Should local definitions in `Context` be polymorphic when `Universe Polymorphism` is on? Currently, it is no.

-----

Depends on #18396 (merged) and #18397.

- [x] Added / updated test-suite
- [ ] ~Corresponding documentation was added / updated~
- [ ] ~Entry added in the changelog~

Overlays:
- smtcoq/smtcoq#141
- LPCIC/coq-elpi#610
- ejgallego/coq-lsp#654 (which includes a discussion on whether it is worth to add a `LetContext` or not)

### CODE REVIEW

**Comment by silene** [comment]
As this is still a draft, I am postponing it to 8.15.

**Comment by herbelin** [comment]
Status of last rebase:
- rather than trying to unify the sematics of Variables/Parameter and Context, we unify only the code, using flags to specialize it to either semantics (i.e. whether class instances are declared, which is done, or whether implicit arguments are used, whether A B:T is the same T or two copies of T, polymorphism of local definitions, which is not done; in particular Equality Scheme in test-suite falis for this reason)
- there is still a syntactic difference (list of assumptions with possibly :> and @{i} for Variables/Parameters all syntax of binders for Context)

@coqbot run full ci

**Comment by herbelin** [comment]
Rebased, including fixing redundant dumpglob in `constrintern.ml` and following the usage of having dumpglob done in `vernacentries.ml`.

Remarks from above [comment](https://github.com/coq/coq/pull/13445#issuecomment-1793508769) still apply.

File: vernac/comAssumption.mli
Comment (ejgallego) [inline vernac/comAssumption.mli:46]: Please document this function and its parameters, maybe also add some more arguments names?

File: vernac/comAssumption.mli
Comment (ejgallego) [inline vernac/comAssumption.mli:70]: Same than the above.

File: vernac/comAssumption.mli
Comment (herbelin) [reply vernac/comAssumption.mli:46]: OK, will do (e.g. I could use the labels `impargs`, `impl` and `univs` as in `declare.ml`).

Also, do you think `declare_local` could/should better go in `declare.ml`? Or a part of it. For instance, in other `declare_foo`, `assumption_message`/`definition_message` and `maybe_declare_manual_implicits` are done in `declare.ml` so maybe should they be done here also in `declare.ml`?

File: vernac/comAssumption.mli
Comment (herbelin) [reply vernac/comAssumption.mli:70]: OK, will do.

**Review by herbelin — COMMENTED** [review]
Thanks, there is indeed room for cleaning the API further.

File: vernac/comAssumption.mli
Comment (ejgallego) [reply vernac/comAssumption.mli:46]: Sure, Hugo, whatever you think is OK; even a very simple note explaining the parameters would suffice IMHO.

Maybe it is only my problem, but I do get a bit confused when functions have so many arguments and I don't have a name for each of them to figure out what they do.

**Comment by proux01** [comment]
@herbelin what is the status of this, can it be reviewed / merged ?

**Comment by herbelin** [comment]
> @herbelin what is the status of this, can it be reviewed / merged ?

I applied the API suggestions of @ejgallego (making the API quite nice and regular I would say) and it apparently needs two other overlays. Otherwise, it is ready to be reviewed.

It currently has quite a lot of atomic commits (all compiling individually), but smashing some of them could also be done if useful.

**Comment by proux01** [comment]
Ok, I'll try to review in the coming weeks.

**Comment by herbelin** [comment]
Note: there are about 10 commits concerned with documenting other files than comAssumption.ml, or adding utilities in other files, or adding tests, which could go to other PRs if ever it might help.

**Comment by proux01** [comment]
Thanks for the offer, I'll tell you while reviewing if ever I'm not confident merging any of those.

**Comment by herbelin** [comment]
@coqbot run full ci

**Comment by herbelin** [comment]
The elpi error is true: I inadvertantly modified the behavior of `interp_context` whose only known client is elpi. I'm waiting for @gares feedback to know if I restore the previous behavior or if he considers that it is better to change elpi.

**Comment by herbelin** [comment]
@coqbot run full ci

File: vernac/comAssumption.ml
Comment (proux01) [inline vernac/comAssumption.ml:119]: Note that (not in this PR), those would probably benefit from using records rather than tuples of tuples.

File: vernac/vernacentries.ml
Comment (proux01) [inline vernac/vernacentries.ml:1501]: We should probably consider (in further PRs) to move the dumpglobs to the com* files. vernacentires should probably be reserved for attributes parsing and dispatching to the various com* files, its already pretty bulky. @SkySkimmer might have a stronger opinion than me here
Meanwhile, regarding the current PR maybe we can just drop the commit " Move dumpglob of Context to vernacentries.ml, like it is already done for Variables." and do the reverse for variables?

File: vernac/comAssumption.ml
Comment (proux01) [inline vernac/comAssumption.ml:20]:
```suggestion
```

**Review by proux01 — APPROVED** [review]
I have a few comments but this mostly LGTM. @herbelin I let you have a look at / address the comments and let's merge then.
It seems at first glance a bit disapointing for a cleaning PR that this isn't decreasing the line count (even taking comments/tests into account) but I hope this is a good move towards making `Variable` and `Context` share a closer beahvior hence imlementation, in future PRs.

File: interp/decls.mli
Comment (herbelin) [reply interp/decls.mli:37]: I'm not sure either. I introduced it to reflect the use of the word `Context`, but we can consider `Context` has not carrying a different semantics than `Let` and `Decls.Let` would be fine also...

File: vernac/comAssumption.ml
Comment (herbelin) [reply vernac/comAssumption.ml:29]: There are different usages about skipping a line or not between a comment and the function, but why not.

File: vernac/comAssumption.ml
Comment (herbelin) [reply vernac/comAssumption.ml:34]: Somehow the question of supporting `#[clearbody] Context ... (a:=t) ...` is also to be solved.

In any case, I'm ready to add it as soon as now if it is clear we want it (with impact to remove the body of all defs of the context, I guess).

File: vernac/declare.ml
Comment (herbelin) [reply vernac/declare.ml:772]: Why not. I guess it is a compromise to find between convenience, separation of concepts, and compatibility. I have no clear idea of what is best.

File: vernac/comAssumption.ml
Comment (herbelin) [reply vernac/comAssumption.ml:110]: Right, that would be clearer.

File: vernac/comAssumption.ml
Comment (herbelin) [reply vernac/comAssumption.ml:117]: OK

File: vernac/comAssumption.ml
Comment (herbelin) [reply vernac/comAssumption.ml:20]: I think I find it clearer to skip a line here.

**Comment by herbelin** [comment]
And thanks a lot for the careful reading.

File: vernac/comAssumption.ml
Comment (proux01) [reply vernac/comAssumption.ml:34]: Let's do that in a separate PR

File: vernac/declare.ml
Comment (proux01) [reply vernac/declare.ml:772]: Having two different ways to do the same thing, one of which does something completely different in another context, seems rather bad to me. I find it more confusing than useful. I agree with the compatibility concern, we should probably do a CI run to assess that. Anyway, thoughts for a further PR.

File: vernac/vernacentries.ml
Comment (proux01) [reply vernac/vernacentries.ml]: Done, I'll merge by the end of the week if there are no more comments.

**Comment by proux01** [comment]
@coqbot run full ci

**Comment by proux01** [comment]
@coqbot merge now

**Pull request merged**
